### PR TITLE
Add missing value splitting (MIA)

### LINKED ISSUE #457: Random forests with missing X-values
Tree-based methods are known for their ability to seamlessly handle missing values in the feature matrix X. We should make GRF able to handle such missingness also. As a first attempt, we plan to implement the MIA method proposed by Twala et al (2008).

At a high level, the MIA approach can be thought of as creating two copies of each feature Xj with missingness: One where the missing values are coded as -infinity, and one where the missing values are coded as +infinity. CART splitting then proceeds as usual.

The above simple implementation, however, is not desirable from a memory-use standpoint, as it involves doubling the number of features. We should thus add to GRF an implementation of MIA splitting that doesn't involve physically duplicating the data.

Twala, B. E. T. H., M. C. Jones, and David J. Hand. "Good methods for coping with missing data in decision trees." <i>Pattern Recognition Letters</i> 29.7 (2008): 950-956.

### PR DESCRIPTION
**Pull request by erikcs**
Add support for missing values in the form of IEEE 754 [NaN](https://en.cppreference.com/w/cpp/numeric/math/nan)s in `X` with MIA splitting (#457). At each split, check whether to send NaNs left or right. Also check splitting on NaN.

- [x] Regression forest 
    *Only done for smallQ. If X contains any NaNs, smallQ will always be used.
    *Adds R test and characterization tests
- [x] Instrumental forest
    *Only done for smallQ. If X contains any NaNs, smallQ will always be used.
    *Adds R test and characterization tests
- [x] Quantile forest
    *Only done for smallQ. If X contains any NaNs, smallQ will always be used.
    *Adds R test and characterization tests

_Details_

* The missing value token is `NaN`. From R either `NaN` or  `NA` can be used

* `Tree` gets a new data structure `vector<bool> send_missing_left` indicating the NaN direction for each node (the default is left). 

* `Tree::find_leaf_node` has updated logic to place NaNs. (In standard conforming C++ all logical operators involving NaNs evaluate to false (i.e. NaN == NaN is false)).

* `TreeTrainer::split_node` has updated logic to push NaN samples to correct node

* `NaN`s are checked for in the O(node.size) sweep done before all splits. If NaNs are encountered an if branch at the beginning of the split loop evaluates sending NaNs in that node L/R by starting off the respective `*_left` running variables with its missing component or zero, if not everything proceeds as normal, the rest of the splitting rules are unchanged.

* `std::stable_sort` is needed for cross platform comparison of results with NaN (std::sort did not break NaN ties the same way and the summation affects float rounding) 

* `Data::::load_from_whitespace_file` needs a simple modification to allow for reading in NaNs from files cross platform

R: 

* input_utilities' validate_X is updated with `allow.na` which is true for all forests except local linear

* `plot` shows the direction NaN is sent with the fill color of the arrow. If a NaN is sent right, the arrow is empty, otherwise black
[MIAplot](https://user-images.githubusercontent.com/7185264/75124148-2d004e80-5662-11ea-8753-109cadb1534b.png)

* Each forest gets a new attribute `has.missing.values`

Checks:

* In addition to added tests, a correctness exercise is [here](https://gist.github.com/erikcs/84958c6b9e48ec62aa9db44cbd948773)

* A C++ unit test that small q / big Q are identical is added [SplittingRuleSmallqBigQTest.cpp](https://github.com/erikcs/grf/blob/MIA/core/test/splitting/SplittingRuleSmallqBigQTest.cpp) (issue #617) 

    _Note:_ Will be removed before merge (A later PR will remove big Q so small Q will always be used) 194135a

* [valgrind_check_MIA.txt](https://gist.github.com/erikcs/0defeecd45f2c0de1226c0ea871337b2)

* [causal_benchmark.R](https://gist.github.com/erikcs/77ff794c1ea323605bb091dffe85ee74)

(closes #457)

### CODE REVIEW

**Comment by swager** [comment]
@imkemayer @julierennes FYI -- we're making progress on implementing MIA forests directly without augmenting the X-matrix. This is still in draft phase, but let's discuss when this is ready.

Also, for context, here are the background references:

Josse, Julie, et al. "On the consistency of supervised learning with missing values." arXiv preprint arXiv:1902.06931 (2019).
Mayer, Imke, et al. "Doubly robust treatment effect estimation with missing attributes." arXiv preprint arXiv:1910.10624 (2019).

File: core/src/commons/Data.cpp [hunk 1]
```diff
@@ -174,22 +175,41 @@ void Data::get_all_values(std::vector<double>& all_values,
...
+  }
 
   for (size_t i = 0; i < samples.size(); i++) {
     sorted_samples[i] = samples[index[i]];
     all_values[i] = get(sorted_samples[i], var);
   }
 
-  all_values.erase(unique(all_values.begin(), all_values.end()), all_values.end());
+  all_values.erase(unique(all_values.begin(), all_values.end(), [&](const double& lhs, const double& rhs) {
+    if (std::isnan(lhs) && std::isnan(rhs)) {
```
Comment (jtibshirani) [inline core/src/commons/Data.cpp]: I think this could be written as
```
return lhs == rhs || (std::isnan(lhs) && std::isnan(rhs))
```

The same suggestion applies to the `stable_sort` call above.

File: core/src/commons/Data.cpp [hunk 2]
```diff
@@ -174,22 +175,41 @@ void Data::get_all_values(std::vector<double>& all_values,
    // fill with [0, 1,..., samples.size() - 1]
   std::iota(index.begin(), index.end(), 0);
   // sort index based on the split values (argsort)
-  std::sort(index.begin(), index.end(), [&](const size_t& lhs, const size_t& rhs)
-    {return all_values[lhs] < all_values[rhs];});
+  if (contains_nan()) {
+    // the NaN comparison places all NaNs at the beginning
+    // this will be slightly slower if there are no NaNs, thus the branch
+    std::stable_sort(index.begin(), index.end(), [&](const size_t& lhs, const size_t& rhs) {
```
Comment (jtibshirani) [inline core/src/commons/Data.cpp]: Did you test the option of alway using `std::stable_sort` and notice a difference in runtime? It looks slower from benchmarks, but not too much slower, and I wonder if we could simplify here and always use `std::stable_sort`. If feels more robust from a cross-platform perspective, as it is a bit of an accident that for non-NaNs, ties are handled the same way in different sort implementations.

**Comment by jtibshirani** [comment]
One last note about tests. First, thanks a lot for adding those new characterization tests. It's great that the old ones still pass without modification -- it shows we didn't inadvertently change the standard behavior.

Some suggestions for C++ behavioral tests:
* In a dataset with missing values, we should be able to negate all the `X` values and still obtain the same results. This helps check left/ right symmetry in how we handle NAs.
* For a variable with no missing values, say we split at value `x`. We should be able make all values on one side of `x` into NAs, and still obtain the same split. (Perhaps this could be written as unit tests for the `SplittingRule` classes?)

File: core/src/commons/Data.cpp (hunk 2 above)
Comment (erikcs) [reply core/src/commons/Data.cpp]: Yes, there is a difference, but not humongous. Will simplify it, as it looks strange from a robustness perspective.

```R
# A large mix continuous/discrete forest (num.threads=12)
set.seed(1)
n <- 1000000
p <- 20
X <- matrix(rnorm(n * p), n, p)
X[1:n, 11:p] <- sample(1:200, n*10, replace = T)
W <- rbinom(n, 1, 0.5)
Y <- pmax(X[, 1], 0) * W + X[, 2] + pmin(X[, 3], 0) + rnorm(n)
system.time(c.forest <- causal_forest(X, Y, W, seed = 1, num.trees = 500))

# master with only small q and std::sort
     user    system   elapsed 
14555.096    62.938  1398.976 

# master with only small q and std::stable_sort
15518.996    62.335  1472.956
```

**Comment by erikcs** [comment]
@imkemayer yes, factor encodings can be handled with the optional package https://github.com/grf-labs/sufrep, which I do not think take missing values into account, @halflearned do you have a suggestion for an appropriate modification to that package? 

Different categories of missing values sounds interesting, that would need a different approach to splitting than this I think.

**Comment by jtibshirani** [comment]
> In a dataset with missing values, we should be able to negate all the X values and still obtain the same results. This helps check left/ right symmetry in how we handle NAs.

@erikcs pointed out offline that this check is not so straightforward. In particular, a forest trained on `X` will partition the data in a symmetric way to a forest trained on `-X`, but the splits themselves won't be identical. One forest would choose a split (-\infinity, b] and (b, \infinity), whereas the other would choose (-\infinity, -c] and (c, \infinity) where b comes right before in the parameter's possible split values. In order to get the test to work we'll need to disable honesty and only check the OOB predictions for equivalence.

File: core/src/splitting/RegressionSplittingRule.cpp [hunk 3]
```diff
@@ -142,36 +160,55 @@ void RegressionSplittingRule::find_best_split_value_small_q(const Data& data,
...
-
-    // Skip this split if one child is too small.
-    if (n_left < min_child_size) {
-      continue;
-    }
-
-    // Stop if the right child is too small.
-    size_t n_right = size_node - n_left;
-    if (n_right < min_child_size) {
+  for (int j = 0; j < 2; ++j) {
```
Comment (jtibshirani) [inline core/src/splitting/RegressionSplittingRule.cpp]: This is a great strategy, but for me the notation was a hard to follow. Here is a suggestion for how to clarify it (assuming it doesn't cause a perf regression):

```
  // Compute decrease of impurity for each possible split
  for (bool send_left : {false, true}) {
    if (send_left)  {
      // A normal split with no NAs, so we can stop early.
      if (n_missing == 0) {
        break;
      }
      
      // It is not necessary to adjust n_right or sum_right as the the missing
      // part is included in the total sum.
      n_left = n_missing;
      weight_sum_left = weight_sum_missing;
      sum_left = sum_missing;
    }
  ....
```

File: core/src/splitting/InstrumentalSplittingRule.cpp [hunk 4]
```diff
@@ -16,6 +16,7 @@
  #-------------------------------------------------------------------------------*/
 
 #include <algorithm>
+#include <cmath> // TODO header includes consistency: this is not needed in RegressionSplitting (std::isnan)
```
[the remark this answers is not in the document]
Comment (erikcs) [reply core/src/splitting/InstrumentalSplittingRule.cpp]: I meant that RegressionSplitting.cpp compiles without including cmath, but not InstrumentalSplittingRule.cpp, because the former has some extra includes in RegressionSplitting.h that InstrumentalSplittingRule.h does not, even thought the required functionality from imports is the same.

I can file a "code quality" issue to do a consistency "checkup" later?

[the remark this answers is not in the document]
Comment (jtibshirani) [reply core/src/splitting/InstrumentalSplittingRule.cpp]: 👍 a new issue sounds good.

**Comment by halflearned** [comment]
Just to respond to the [comment above](https://github.com/grf-labs/grf/pull/612#issuecomment-590167224) on our [sufficient representations](https://github.com/grf-labs/sufrep) package, it unfortunately currently cannot handle the case of missing values. That is, unless we manually let code NA as a factor _and_ assume enough that we can treat it as any category (likely MCAR should do it). But thanks for letting me know that this could be a useful feature!

**Comment by erikcs** [comment]
> I reviewed everything and left a few comments on the splitting rule changes. Here are the other items I'm aware of to be addressed before merging:
> 
> * Add one or two invariant tests ([#612 (comment)](https://github.com/grf-labs/grf/pull/612#issuecomment-590140382)).

@jtibshirani please have a look at the test for the second invariance, added in 0646630

It checks that setting all the values to the left of a full data split to missing, yields the same split. The behavior is not symmetric as setting all the values to the right of a full data split to missing will change the partitions, a bigger impurity decrease might be obtained by having all the missing on the left together with some non-missing.

The first invariance is tricky, I do not know how to test this exactly. Since the summation order in the impurity calculations will be reversed, rounding error may accumulate differently and even though b comes right before c in the parameter's possible split values, a different split will sometimes occur.

File: core/src/commons/Data.cpp (hunk 2 above)
Comment (erikcs) [reply core/src/commons/Data.cpp]: Just adding a post merge note that the main (minor) slowdown in causal_benchmark.R arises from the change to always use stable sorting:

```
MIA @ ce2cbeb (just before always use stable sort)

         oob               test              stdratio
Y.error   "0.048 +/- 0.001" "0.049 +/- 0.001" "2.667 +/- 0.110"
tau.error "0.010 +/- 0.001" "0.009 +/- 0.001" "1.298 +/- 0.093"
              user.self          sys.self          elapsed
Y.time*        "16.586 +/- 0.067" "0.070 +/- 0.011" "16.677 +/- 0.078"
Y.time.pred   "2.486 +/- 0.017"  "0.051 +/- 0.007" "2.545 +/- 0.018"
Y.time.ci     "5.432 +/- 0.049"  "0.050 +/- 0.005" "5.494 +/- 0.055"
tau.time*      "15.723 +/- 0.055" "0.042 +/- 0.004" "15.790 +/- 0.056"
tau.time.pred "2.531 +/- 0.039"  "0.029 +/- 0.004" "2.565 +/- 0.040"
tau.time.ci   "4.841 +/- 0.073"  "0.036 +/- 0.004" "4.890 +/- 0.077"
```

```
MIA @ f948d4b (always use stable sort)

          oob               test              stdratio
Y.error   "0.048 +/- 0.001" "0.049 +/- 0.001" "2.667 +/- 0.110"
tau.error "0.010 +/- 0.001" "0.009 +/- 0.001" "1.298 +/- 0.093"
              user.self          sys.self          elapsed
Y.time*        "17.205 +/- 0.048" "0.075 +/- 0.010" "17.302 +/- 0.058"
Y.time.pred   "2.492 +/- 0.023"  "0.055 +/- 0.006" "2.550 +/- 0.025"
Y.time.ci     "5.485 +/- 0.038"  "0.054 +/- 0.004" "5.546 +/- 0.039"
tau.time*      "16.035 +/- 0.071" "0.048 +/- 0.004" "16.106 +/- 0.075"
tau.time.pred "2.473 +/- 0.023"  "0.027 +/- 0.003" "2.502 +/- 0.024"
tau.time.ci   "4.775 +/- 0.076"  "0.028 +/- 0.002" "4.808 +/- 0.078"
```

```
MIA @ latest

          oob               test              stdratio
Y.error   "0.048 +/- 0.001" "0.049 +/- 0.001" "2.667 +/- 0.110"
tau.error "0.010 +/- 0.001" "0.009 +/- 0.001" "1.298 +/- 0.093"
              user.self          sys.self          elapsed
Y.time*        "17.259 +/- 0.078" "0.073 +/- 0.010" "17.369 +/- 0.101"
Y.time.pred   "2.519 +/- 0.024"  "0.055 +/- 0.007" "2.583 +/- 0.024"
Y.time.ci     "5.510 +/- 0.064"  "0.055 +/- 0.005" "5.574 +/- 0.066"
tau.time*      "16.049 +/- 0.079" "0.051 +/- 0.008" "16.139 +/- 0.095"
tau.time.pred "2.555 +/- 0.051"  "0.025 +/- 0.003" "2.582 +/- 0.053"
tau.time.ci   "4.801 +/- 0.048"  "0.029 +/- 0.002" "4.834 +/- 0.050"
```

The [causal_benchmark.txt](https://gist.github.com/erikcs/77ff794c1ea323605bb091dffe85ee74) linked to in in this issue summary contained a wrong entry, sorry.
